src/block/block_compact.c

```diff
@@ -65,6 +70,295 @@ __wt_block_compact_get_progress_stats(WT_SESSION_IMPL *session, WT_BM *bm,
+ *     Estimate how much more work the compaction needs to do for the given file.
+ */
+static void
+__block_compact_estimate_remaining_work(WT_SESSION_IMPL *session, WT_BLOCK *block)
+{
+    WT_EXT *ext;
+    wt_off_t avg_block_size, avg_internal_block_size, depth1_size, leaves_per_internal_page;
+    wt_off_t compact_start_off, extra_space, file_size, last, off, rewrite_size, size, write_off;
+    uint64_t n, pages_to_move, pages_to_move_orig, total_pages_to_move;
+    int compact_pct_tenths, iteration;
+    bool skip;
+
+    /*
+     * We must have reviewed at least some interesting number of pages for any estimates below to be
+     * worthwhile.
+     */
+    if (block->compact_pages_reviewed < WT_THOUSAND)
+        return;
+
+    /* Assume that we have already checked whether this file can be skipped. */
+    WT_ASSERT(session, block->compact_pct_tenths > 0);
+
+    /*
+     * Get the average block size that we encountered so far during compaction. Note that we are not
+     * currently accounting for overflow pages, as compact does not currently account for them
+     * either.
+     */
+    avg_block_size = (wt_off_t)WT_ALIGN(
+      block->compact_bytes_reviewed / block->compact_pages_reviewed, block->allocsize);
+
+    /* We don't currently have a way to track the internal page size, but this should be okay. */
+    avg_internal_block_size = 4096;
+    leaves_per_internal_page =
+      (wt_off_t)(block->compact_pages_reviewed / block->compact_internal_pages_reviewed);
+    depth1_size = avg_block_size * leaves_per_internal_page + avg_internal_block_size;
+
+    __wt_verbose_debug2(session, WT_VERB_COMPACT,
+      "%s: the average block size is %" PRId64 " bytes (based on %" PRIu64 " blocks)", block->name,
+      avg_block_size, block->compact_pages_reviewed);
+    __wt_verbose_debug2(session, WT_VERB_COMPACT, "%s: reviewed %" PRIu64 " internal pages so far",
+      block->name, block->compact_internal_pages_reviewed);
+
+    /*
+     * We would like to estimate how much data will be moved during compaction, so that we can
+     * inform users how far along we are in the process. We will estimate how many pages are in the
+     * last part of the file (typically the last 10%) and where they will be written using the
+     * first-fit allocation, and then repeat as long as we continue to make progress, to emulate the
+     * behavior of the actual compaction. This does not account for all complexities that we may
+     * encounter, but the hope is that the estimate would be still good enough.
+     */
+
+    __wt_spin_lock(session, &block->live_lock);
+
+    compact_pct_tenths = block->compact_pct_tenths;
+    extra_space = 0;
+    file_size = block->size;
+    pages_to_move = 0;
+    total_pages_to_move = 0;
+    write_off = 0;
+
+#define WT_EXT_SIZE_TO_LEAF_PAGES(ext_size)                          \
+    (uint64_t)((ext_size) / depth1_size * leaves_per_internal_page + \
+      ((ext_size) % depth1_size) / avg_block_size)
+
```

Comment:
An explanatory comment would be helpful here. Using a more descriptive variable instead of depth1_size could help as well.